### PR TITLE
Persistent Node Ids

### PR DESCRIPTION
**Pull request by ywelsch**
Nodes are currently identified in the cluster by IDs that are randomly generated during node startup. That means they change every time the node is restarted (i.e. they correspond to process ids). While this doesn't matter for ES proper, it makes it hard for external services to track nodes. Persistent node ids ensure that the same id can be reused across restarts, facilitating tracking of nodes in node stats / cluster state APIs.

In contrast to #17811 (which this PR supersedes), this PR keeps process ids around for the code that relies on its semantics (mainly discovery-related parts).

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 1]
```diff
@@ -80,7 +88,8 @@ public static boolean isIngestNode(Settings settings) {
     }
 
     private final String nodeName;
-    private final String nodeId;
+    private final String processId;
+    private final String persistentNodeId;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: Can we just get away with calling it `nodeId`? We don't have non-persistent node IDs. Yes it's slightly confusing in the transition from what use to be `nodeId` and is now `processId`, but I think in the long-run it would be better if we just called it `nodeId`?

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 2]
```diff
@@ -332,6 +408,7 @@ public String toString() {
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(getId());
         builder.field("name", getName());
+        builder.field("persistent_id", getPersistentNodeId());
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: This is especially a place where I think just `node_id` would be better?

File: core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeTests.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.elasticsearch.test.VersionUtils.randomVersion;
+
+public class DiscoveryNodeTests extends ESTestCase {
+
+    private static final AtomicInteger idGenerator = new AtomicInteger();
+
+    /**
+     * guarantees to generate a node that is unique for all aspects but for it's roles and version (which are chosen randomly)
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeTests.java]: Nit: `it's` -> `its`

File: core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeTests.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    DiscoveryNode randomNode() {
+        Map<String, String> attributes = new HashMap<>();
+        for (int attr = (idGenerator.get() > 0 ? 1 : 0) + randomInt(3); attr > 0; attr--) {
+            attributes.put("attr_" + attr, "v_" + idGenerator.incrementAndGet());
+        }
+        final DiscoveryNode node = new DiscoveryNode(
+            "name_" + idGenerator.incrementAndGet(),
+            "processId_" + idGenerator.incrementAndGet(),
+            "persistentNodeId_" + idGenerator.incrementAndGet(),
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodeTests.java]: `persistentNodeId` -> `nodeId_` :smile:

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 5]
```diff
@@ -140,6 +140,16 @@
...
         Setting.boolSetting("node.ingest", true, Property.NodeScope);
+
+    /**
+     * controls whether the node is allowed to persist things like metadata to disk
+     * Note that this does not control whether the node stores actual indices (see
+     * {@link #NODE_DATA_SETTING}). However, if this is false, {@link #NODE_DATA_SETTING}
+     * and {@link #NODE_MASTER_SETTING} must also be false.
+     *
+     */
+    public static final Setting<Boolean> NODE_LOCAL_STORAGE_SETTING = Setting.boolSetting("node.local_storage", true, Property.NodeScope);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I do not think we should have this setting. I think that for the case of tribe nodes, the sub-nodes can just use the node ID of the "parent" mixed (in a repeatable way, e.g., as a seed to the random base-64 UUID) with a combination of the tribe node name (the `t1` part of `tribe.t1.` in the settings) and the downstream cluster name.

**Comment by jasontedor** [comment]
@ywelsch I've given it a first pass and left some comments regarding the node local storage setting and just calling it node ID instead of persistent node ID. I'll do a more thorough review after those are addressed (especially the latter one since it's going to touch a lot of places).

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 5 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/node/Node.java]: In e57232f I've made it so the node id of the tribe client nodes is based on the node id of the parent and the tribe node name. I've still kept the `NODE_LOCAL_STORAGE_SETTING` flag so that the tribe client nodes don't persist this name on disk (as they share the same data path with their tribe parent node). The tribe parent node generates and persists its node id in the same way as any other kind of node.

**Comment by ywelsch** [comment]
@jasontedor I've added support for choosing a node id for tribe node clients in a deterministic way and renamed the persistentNodeId fields. Please have another look.

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java [hunk 6]
```diff
@@ -584,6 +606,26 @@ public Builder localNodeId(String localNodeId) {
             return this;
         }
 
+        public DiscoveryNode nodeWithSameAddress(DiscoveryNode node) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: I have no idea how often this is called but maybe it's worth to add the node ID as a key in the map too? ie. put multiple keys for one `DiscoveryNode` into the `nodes` map?

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 7]
```diff
@@ -128,23 +133,30 @@ public String toString() {
...
-     * If true automatically append node id to custom data paths.
+     * If true automatically append node lock id to custom data paths.
      */
-    public static final Setting<Boolean> ADD_NODE_ID_TO_CUSTOM_PATH =
-        Setting.boolSetting("node.add_id_to_custom_path", true, Property.NodeScope);
+    public static final Setting<Boolean> ADD_NODE_LOCK_ID_TO_CUSTOM_PATH =
+        Setting.boolSetting("node.add_lock_id_to_custom_path", true, Property.NodeScope);
+
+
+    public static final Setting<Long> NODE_ID_SEED_SETTING =
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: can you document this setting please I always struggle to figure it out :)

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 8]
```diff
@@ -641,6 +674,17 @@ public boolean hasNodeFile() {
     }
 
     /**
+     * returns the unique uuid describing this node. The uuid is persisted in the data folder of this node
+     * and is reused across restarts.
+     **/
+    public String nodeId() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: can we make this a `getNodeId()`

File: core/src/main/java/org/elasticsearch/env/NodeMetaData.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.gateway.MetaDataStateFormat;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Objects;
+
+/**
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/env/NodeMetaData.java]: give us some chars that explain what this is and where it's used?

File: core/src/main/java/org/elasticsearch/env/NodeMetaData.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+ */
+public final class NodeMetaData {
+
+    private static final String NODE_ID_KEY = "node_id";
+
+    private final String nodeId;
+
+    public NodeMetaData(final String nodeId) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/env/NodeMetaData.java]: does this need to be public?

File: core/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java [hunk 11]
```diff
@@ -182,9 +181,14 @@ public final T read(Path file) throws IOException {
             try (final IndexInput indexInput = dir.openInput(file.getFileName().toString(), IOContext.DEFAULT)) {
                  // We checksum the entire file before we even go and parse it. If it's corrupted we barf right here.
                 CodecUtil.checksumEntireFile(indexInput);
-                CodecUtil.checkHeader(indexInput, STATE_FILE_CODEC, STATE_FILE_VERSION, STATE_FILE_VERSION);
+                final int fileVersion = CodecUtil.checkHeader(indexInput, STATE_FILE_CODEC, MIN_COMPATIBLE_STATE_FILE_VERSION,
+                    STATE_FILE_VERSION);
                 final XContentType xContentType = XContentType.values()[indexInput.readInt()];
-                indexInput.readLong(); // version currently unused
+                if (fileVersion == 0) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java]: can you compare this against  `MIN_COMPATIBLE_STATE_FILE_VERSION` instead? and maybe we should add a some comment when we can remove it?

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 12]
```diff
@@ -322,8 +332,9 @@ public Node start() {
 
         validateNodeBeforeAcceptingRequests(settings, transportService.boundAddress());
 
-        DiscoveryNode localNode = injector.getInstance(DiscoveryNodeService.class)
-                .buildLocalNode(transportService.boundAddress().publishAddress());
+        String nodeId = injector.getInstance(NodeEnvironment.class).nodeId();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: Can we just assign `NodeEnvironment` next to `Environment` and don't do the guice lookup?

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 5 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: hmm can we just do the same as we do in here `DiscoveryNode.nodeRequiresLocalStorage(settings)` to figure this out?

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java (hunk 6 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: It is only called when a node that joins the cluster is added on the master node. In test code it is also asserted as an invariant on `DiscoveryNodes`. As the number of nodes is limited I prefer keeping it out of the `nodes` map for now.

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java (hunk 7 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: addressed in 7dbe1bf

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java (hunk 8 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: ok, done in 7dbe1bf

File: core/src/main/java/org/elasticsearch/env/NodeMetaData.java (hunk 9 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/env/NodeMetaData.java]: done in 7dbe1bf

File: core/src/main/java/org/elasticsearch/env/NodeMetaData.java (hunk 10 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/env/NodeMetaData.java]: nope, made it package-private

File: core/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java (hunk 11 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java]: I have made it more explicit what we're trying to achieve. Have a look at 7dbe1bf.

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 12 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/node/Node.java]: `NodeEnvironment` is created in the constructor and `localNode` injection happens in the `start` method. I could make `nodeEnvironment` or `nodeId` a final field in the `Node` class, but I'm not sure that makes the code much nicer.

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 5 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I don't know how to create the setting so that validation fails if `node.master` or `node.data` is set. The validation logic has only access to the new value of the setting and not the settings object.

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java (hunk 6 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: cool

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 5 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I am not sure I am following :)

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/node/Node.java]: me neither :D Can you reformulate your comment / question? Is it about removing this setting?

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I discussed this with @ywelsch and we agreed that we keep it for now since it complicates things otherwise

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/node/Node.java]: As discussed I've made this a filtered setting.

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 13]
```diff
@@ -97,14 +105,15 @@ public static boolean isIngestNode(Settings settings) {
...
-     * @param version    the version of the node.
+     * @param id               the nodes unique (ephemeral and persistent) node id
+     * @param address          the nodes transport address
+     * @param attributes       node attributes
+     * @param roles            node roles
+     * @param version          the version of the node
      */
-    public DiscoveryNode(String nodeId, TransportAddress address, Map<String, String> attributes, Set<Role> roles, Version version) {
-        this("", nodeId, address.getHost(), address.getAddress(), address, attributes, roles, version);
+    public DiscoveryNode(String id, TransportAddress address, Map<String, String> attributes, Set<Role> roles,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: hmm. I think this is tricky having no id string. How about using the supplied id as the nodeId and generate a random ephemeral one?

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 14]
```diff
@@ -238,6 +251,13 @@ public String getId() {
     }
 
     /**
+     * The unique ephemeral id of the node.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: can we add some docs as to what we mean with ephemeral - most notably when it changes.

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 15]
```diff
@@ -313,15 +365,10 @@ public String toString() {
...
-        if (nodeId != null) {
-            sb.append('{').append(nodeId).append('}');
-        }
-        if (Strings.hasLength(hostName)) {
-            sb.append('{').append(hostName).append('}');
-        }
-        if (address != null) {
-            sb.append('{').append(address).append('}');
-        }
+        sb.append('{').append(nodeId).append('}');
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: why did you change this?

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java [hunk 16]
```diff
@@ -279,53 +305,69 @@ public DiscoveryNode resolveNode(String node) {
                 } else if (nodeExists(nodeId)) {
                     resolvedNodesIds.add(nodeId);
                 } else {
-                    // not a node id, try and search by name
+                    boolean matchedEphemeralId = false;
                     for (DiscoveryNode node : this) {
-                        if (Regex.simpleMatch(nodeId, node.getName())) {
+                        if (node.getEphemeralId().equals(nodeId)) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: why did add this extra resolveNode path? isn't ephemeralId supposed to be internal and shouldn't be used in an API?

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java (hunk 15 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: nodeId is never null
ephemeralId is never null
hostName/address is never null

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java (hunk 16 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: sure, I can remove it.

File: core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java (hunk 13 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: I don't understand what you mean here. This constructor is mainly used by tests where we don't care much about the actual values and construction of temporary nodes for pinging. I don't think we should just randomly generate a ephemeral id here. Better would be to get rid of this constructor altogether...

File: core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java [hunk 17]
```diff
@@ -980,7 +980,10 @@ public void run() {
         public void onNodeAck(DiscoveryNode node, @Nullable Throwable t) {
             if (!ackedTaskListener.mustAck(node)) {
                 //we always wait for the master ack anyway
-                if (!node.equals(nodes.getMasterNode())) {
+                DiscoveryNode masterNode = nodes.getMasterNode();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java]: what was the rational behind this change? equals checks for all of these no?

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java]: previously, equals did only check for equality of ephemeral id. Extending this check now to all attributes felt:
1) inefficient (always compare all attributes just to detect that it's same node)
2) unsafe w.r.t. future extensions (where we could for example make it possible for node to have attributes changing at runtime).

File: core/src/main/java/org/elasticsearch/tribe/TribeService.java [hunk 18]
```diff
@@ -241,6 +246,10 @@ static Settings buildClientSettings(String tribeName, Settings globalSettings, S
         sb.put(Node.NODE_DATA_SETTING.getKey(), false);
         sb.put(Node.NODE_MASTER_SETTING.getKey(), false);
         sb.put(Node.NODE_INGEST_SETTING.getKey(), false);
+        // node id of a tribe client node is determined by node id of parent node and tribe name
+        int nodeIdSeed = StringHelper.murmurhash3_x86_32(new BytesRef(parentNodeId + "/" + tribeName), 0);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/tribe/TribeService.java]: +1. nice solution.

**Comment by ywelsch** [comment]
superseded by #19140
